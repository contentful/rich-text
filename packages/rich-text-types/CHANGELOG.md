# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [15.3.6](https://github.com/contentful/rich-text/compare/v15.3.5...v15.3.6) (2021-09-15)


### Bug Fixes

* allow table-row type to have both cells ([#267](https://github.com/contentful/rich-text/issues/267)) ([ec40598](https://github.com/contentful/rich-text/commit/ec405982109cb1c16e7adf71a541a98270d7f45b))





## [15.3.5](https://github.com/contentful/rich-text/compare/v15.3.4...v15.3.5) (2021-09-13)


### Bug Fixes

* **rich-text-types:** forbid Tables inside ListItem ([#266](https://github.com/contentful/rich-text/issues/266)) ([fc338bf](https://github.com/contentful/rich-text/commit/fc338bf040b8718057717d2681f800d5e26ba59d))





## [15.3.3](https://github.com/contentful/rich-text/compare/v15.3.2...v15.3.3) (2021-09-07)

**Note:** Version bump only for package @contentful/rich-text-types





## [15.3.2](https://github.com/contentful/rich-text/compare/v15.3.1...v15.3.2) (2021-09-07)

**Note:** Version bump only for package @contentful/rich-text-types





## [15.3.1](https://github.com/contentful/rich-text/compare/v15.3.0...v15.3.1) (2021-09-07)

**Note:** Version bump only for package @contentful/rich-text-types





# [15.3.0](https://github.com/contentful/rich-text/compare/v15.2.0...v15.3.0) (2021-09-06)


### Bug Fixes

* allow the first item in Table to be a TableHeaderRow ([9a6de55](https://github.com/contentful/rich-text/commit/9a6de55ed97ca413d3922958850b8215922c06b5))
* don't export the helper TableHeaderRow type ([179c3cb](https://github.com/contentful/rich-text/commit/179c3cb10c0f65725297445afe955f0cca135005))
* typo ([2282665](https://github.com/contentful/rich-text/commit/2282665924404771b3353dd9b380d8863f1d4c41))


### Features

* add TableHeaderCell type ([5e25ac9](https://github.com/contentful/rich-text/commit/5e25ac9f35be2ad6c7ad7857cbde808cbc3437f9))


# [15.2.0](https://github.com/contentful/rich-text/compare/v15.2.0...v15.1.0) (2021-09-06)


### Bug Fixes

* allow the first item in Table to be a TableHeaderRow ([9a6de55](https://github.com/contentful/rich-text/commit/9a6de55ed97ca413d3922958850b8215922c06b5))
* don't export the helper TableHeaderRow type ([179c3cb](https://github.com/contentful/rich-text/commit/179c3cb10c0f65725297445afe955f0cca135005))
* typo ([2282665](https://github.com/contentful/rich-text/commit/2282665924404771b3353dd9b380d8863f1d4c41))


### Features

* add TableHeaderCell type ([5e25ac9](https://github.com/contentful/rich-text/commit/5e25ac9f35be2ad6c7ad7857cbde808cbc3437f9))



# [15.1.0](https://github.com/contentful/rich-text/compare/v15.0.0...v15.1.0) (2021-08-02)


### Features

* add rowspan ([e0264fb](https://github.com/contentful/rich-text/commit/e0264fbe4d9467eedf78b7c5c3a9825a584124cf))



# [15.0.0](https://github.com/contentful/rich-text/compare/v14.2.0...v15.0.0) (2021-06-15)


### Features

* add table markup support ([c6ae127](https://github.com/contentful/rich-text/commit/c6ae127aa460f4cd26cd6b671cd43fd2714bc650))



## [14.1.2](https://github.com/contentful/rich-text/compare/v14.0.1...v14.1.2) (2020-11-02)


### Bug Fixes

* üêõ configure rollup copy correctly ([3446a33](https://github.com/contentful/rich-text/commit/3446a33fc5711087095a58b088828dfe6066bc7f))



## [14.0.1](https://github.com/contentful/rich-text/compare/v14.0.0...v14.0.1) (2020-01-30)


### Features

* üé∏ emptyDoc representing an empty RT document ([80315ef](https://github.com/contentful/rich-text/commit/80315ef5130b79336336ba31de8a55e42bafe319))



# [14.0.0](https://github.com/contentful/rich-text/compare/v13.4.0...v14.0.0) (2020-01-28)



# [13.4.0](https://github.com/contentful/rich-text/compare/v13.3.0...v13.4.0) (2019-08-01)



# [13.1.0](https://github.com/contentful/rich-text/compare/v13.0.1...v13.1.0) (2019-03-04)



# [13.0.0](https://github.com/contentful/rich-text/compare/v12.2.1...v13.0.0) (2019-01-22)



## [12.1.2](https://github.com/contentful/rich-text/compare/v12.1.1...v12.1.2) (2018-12-14)



## [12.0.1](https://github.com/contentful/rich-text/compare/v12.0.0...v12.0.1) (2018-12-04)


### Bug Fixes

* üêõ add Object.values and Array.prototype.includes polyfills ([9e2ffb4](https://github.com/contentful/rich-text/commit/9e2ffb46e3564a523c74504270b29ccc8b8249ad))



# [12.0.0](https://github.com/contentful/rich-text/compare/v11.0.0...v12.0.0) (2018-11-29)



# [11.0.0](https://github.com/contentful/rich-text/compare/v10.3.0...v11.0.0) (2018-11-27)


### Bug Fixes

* üêõ Removes obsolete field `title` from Hyperlinks ([8cb9027](https://github.com/contentful/rich-text/commit/8cb90279f5348a7fb59f211c2ba209a28fd432be))



# [10.3.0](https://github.com/contentful/rich-text/compare/v10.2.0...v10.3.0) (2018-11-26)



# [10.2.0](https://github.com/contentful/rich-text/compare/v10.1.0...v10.2.0) (2018-11-19)


### Features

* üé∏ Add rich-text-references ([363b4e5](https://github.com/contentful/rich-text/commit/363b4e509e94af0932fd7cece8e56beafe8d67d2))



# [10.1.0](https://github.com/contentful/rich-text/compare/v10.0.5...v10.1.0) (2018-11-16)


### Features

* üé∏ adds json schema generation to rich text types ([8916140](https://github.com/contentful/rich-text/commit/89161404eb911f126be23ba2a146ebf748f7489e))
* üé∏ introduces top-level-block type ([a6bf35e](https://github.com/contentful/rich-text/commit/a6bf35e7c9ca35915a512de774b3a3fdc4c76e5d))



## [10.0.1](https://github.com/contentful/rich-text/compare/v10.0.0...v10.0.1) (2018-11-08)



# [10.0.0](https://github.com/contentful/rich-text/compare/v9.0.2...v10.0.0) (2018-11-02)


### Features

* üé∏ removes nodeClass from Node interface ([09b8162](https://github.com/contentful/rich-text/commit/09b8162bcba65bc13afa14b2b5ff046c9fed7b3b))


### BREAKING CHANGES

* Removes accidentally added nodeClass



## [9.0.2](https://github.com/contentful/rich-text/compare/v9.0.1...v9.0.2) (2018-10-31)



# [9.0.0](https://github.com/contentful/rich-text/compare/v8.0.3...v9.0.0) (2018-10-30)


### Features

* üé∏ Explicitly declare nodeClass and nodeType in core types ([0749c61](https://github.com/contentful/rich-text/commit/0749c6199bb2681509608539c76bd8149cade564))


### BREAKING CHANGES

* Potentially breaks TypeScript libraries pulling this in as a dependency
if they are not explicitly stating nodeClass and nodeType.



## [8.0.3](https://github.com/contentful/rich-text/compare/v8.0.2...v8.0.3) (2018-10-30)



## [8.0.2](https://github.com/contentful/rich-text/compare/v8.0.1...v8.0.2) (2018-10-29)



## [8.0.1](https://github.com/contentful/rich-text/compare/v8.0.0...v8.0.1) (2018-10-26)


### Bug Fixes

* üêõ Revert mock hypenation commits ([bde9432](https://github.com/contentful/rich-text/commit/bde94323fcc02bf5ab3feeef46a9d8fc8b08d59b))


### Features

* üé∏ take all packages out of "demo mode" ([815f18b](https://github.com/contentful/rich-text/commit/815f18be6a914e7e4782790ee46053689712494b))


### BREAKING CHANGES

* renames all packages



# 6.0.0 (2018-10-24)
