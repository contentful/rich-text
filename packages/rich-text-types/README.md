# rich-text-types

Type definitions and constants for the Contentful rich text field type.
