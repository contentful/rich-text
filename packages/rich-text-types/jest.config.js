/* eslint-disable */

const getBaseConfig = require('../../baseJestConfig');
const packageJson = require('./package.json');
const packageName = packageJson.name.split('@contentful/')[1];

module.exports = {
  ...getBaseConfig(packageName),
};
