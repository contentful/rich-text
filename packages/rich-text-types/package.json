{
  "name": "@contentful/rich-text-types",
  "version": "15.15.1",
  "main": "dist/index.js",
  "typings": "dist/types/index.d.ts",
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/contentful/rich-text.git"
  },
  "license": "MIT",
  "engines": {
    "node": ">=6.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "npm run generate-json-schema && tsc --module commonjs && rollup -c rollup.config.js",
    "start": "tsc && rollup -c rollup.config.js -w",
    "lint": "tslint -t codeFrame '@(src|bin)/*.ts'",
    "generate-json-schema": "ts-node -O '{\"module\": \"commonjs\"}' ./tools/jsonSchemaGen"
  },
  "devDependencies": {
    "@cspotcode/source-map-consumer": "^0.8.0",
    "@types/jest": "^27.0.1",
    "@types/node": "^14.17.14",
    "faker": "^4.1.0",
    "jest": "^27.1.0",
    "rimraf": "^2.6.3",
    "rollup": "^1.32.1",
    "rollup-plugin-commonjs": "^9.3.4",
    "rollup-plugin-copy": "^2.0.0",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^4.2.4",
    "rollup-plugin-sourcemaps": "^0.6.3",
    "rollup-plugin-typescript2": "^0.30.0",
    "ts-jest": "^27.0.5",
    "ts-node": "^10.2.1",
    "tslib": "^2.3.1",
    "tslint": "^6.1.3",
    "typescript": "^4.4.2",
    "typescript-json-schema": "^0.50.1"
  },
  "gitHead": "c7e34bc0f0172728999b121af43df88995eec305"
}
