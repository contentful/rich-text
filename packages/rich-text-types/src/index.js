"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var blocks_1 = require("./blocks");
exports.BLOCKS = blocks_1.default;
var inlines_1 = require("./inlines");
exports.INLINES = inlines_1.default;
var marks_1 = require("./marks");
exports.MARKS = marks_1.default;
__export(require("./schemaConstraints"));
//# sourceMappingURL=index.js.map