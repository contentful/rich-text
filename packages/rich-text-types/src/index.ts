export { BLOCKS } from './blocks';
export { INLINES } from './inlines';
export { MARKS } from './marks';

export * from './schemaConstraints';

export * from './types';
export * from './nodeTypes';

export { EMPTY_DOCUMENT } from './emptyDocument';

import * as helpers from './helpers';
export { helpers };

export { getSchemaWithNodeType } from './schemas';
