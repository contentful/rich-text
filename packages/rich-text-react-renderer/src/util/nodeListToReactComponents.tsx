import React, { ReactNode } from 'react';
import { helpers, Mark } from '@contentful/rich-text-types';
import { CommonNode, Options } from '..';
import { appendKeyToValidElement } from './appendKeyToValidElement';

export function nodeListToReactComponents(nodes: CommonNode[], options: Options): ReactNode {
  return nodes.map((node: CommonNode, index: number): ReactNode => {
    return appendKeyToValidElement(nodeToReactComponent(node, options), index);
  });
}

export function nodeToReactComponent(node: CommonNode, options: Options): ReactNode {
  const { renderNode, renderMark, renderText, preserveWhitespace } = options;

  if (helpers.isText(node)) {
    let nodeValue: ReactNode = renderText ? renderText(node.value) : node.value;

    if (preserveWhitespace) {
      // Preserve multiple spaces.
      nodeValue = (nodeValue as string).replace(/ {2,}/g, (match) => '\u00A0'.repeat(match.length));

      // Preserve line breaks.
      let lines = (nodeValue as string).split('\n');
      let jsxLines: (string | JSX.Element)[] = [];
      lines.forEach((line, index) => {
        jsxLines.push(line);
        if (index !== lines.length - 1) {
          jsxLines.push(<br />);
        }
      });
      nodeValue = jsxLines;
    }

    return node.marks.reduce((value: ReactNode, mark: Mark): ReactNode => {
      if (!renderMark[mark.type]) {
        return value;
      }
      return renderMark[mark.type](value);
    }, nodeValue);
  } else {
    const children: ReactNode = nodeListToReactComponents(node.content, options);
    if (!node.nodeType || !renderNode[node.nodeType]) {
      return <>{children}</>;
    }
    return renderNode[node.nodeType](node, children);
  }
}
