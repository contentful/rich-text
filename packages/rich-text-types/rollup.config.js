import copy from 'rollup-plugin-copy';

import config from '../../rollup.config';
import { main as outputFile } from './package.json';

const customConfig = () => {
  const baseConfig = config(outputFile);

  return {
    ...baseConfig,
    output: {
      // Maintain the same file names & folders structure as "src"
      // This is necessary to resolve generated JSON schema files
      //
      // https://rollupjs.org/guide/en/#outputpreservemodules
      preserveModules: true,
    },
    plugins: [
      ...baseConfig.plugins,
      copy({
        targets: [
          {
            src: 'src/schemas/generated',
            dest: 'dist/schemas',
          },
        ],
      }),
    ],
  };
};

export default customConfig();
