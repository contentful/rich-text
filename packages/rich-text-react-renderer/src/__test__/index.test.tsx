import React, { ReactNode, ReactNodeArray } from 'react';
import { BLOCKS, Document, INLINES, MARKS } from '@contentful/rich-text-types';

import { CommonNode, documentToReactComponents, Options } from '..';

import {
  embeddedEntryDoc,
  headingDoc,
  hrDoc,
  hyperlinkDoc,
  inlineEntityDoc,
  invalidMarksDoc,
  invalidTypeDoc,
  marksDoc,
  multiMarkDoc,
  olDoc,
  paragraphDoc,
  quoteDoc,
  ulDoc,
} from './documents';
import Paragraph from './components/Paragraph';
import Strong from './components/Strong';
import { appendKeyToValidElement } from '../util/appendKeyToValidElement';
import { nodeListToReactComponents, nodeToReactComponent } from '../util/nodeListToReactComponents';

describe('documentToReactComponents', () => {
  it('returns an empty array when given an empty document', () => {
    const document: Document = {
      nodeType: BLOCKS.DOCUMENT,
      data: {},
      content: [],
    };

    expect(documentToReactComponents(document)).toEqual([]);
  });

  it('renders nodes with default node renderer', () => {
    const docs: Document[] = [
      paragraphDoc,
      headingDoc(BLOCKS.HEADING_1),
      headingDoc(BLOCKS.HEADING_2),
    ];

    docs.forEach(doc => {
      expect(documentToReactComponents(doc)).toMatchSnapshot();
    });
  });

  it('renders marks with default mark renderer', () => {
    const docs: Document[] = [
      marksDoc(MARKS.ITALIC),
      marksDoc(MARKS.BOLD),
      marksDoc(MARKS.UNDERLINE),
      marksDoc(MARKS.CODE),
    ];

    docs.forEach(doc => {
      expect(documentToReactComponents(doc)).toMatchSnapshot();
    });
  });

  it('renders unaltered text with default text renderer', () => {
    const document: Document = paragraphDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders multiple marks with default mark renderer', () => {
    const doc: Document = multiMarkDoc();
    expect(documentToReactComponents(doc)).toMatchSnapshot();
  });

  it('renders nodes with passed custom node renderer', () => {
    const options: Options = {
      renderNode: {
        [BLOCKS.PARAGRAPH]: (node, children) => <Paragraph>{children}</Paragraph>,
      },
    };
    const document: Document = quoteDoc;
    expect(documentToReactComponents(document, options)).toMatchSnapshot();
  });

  it('renders marks with the passed custom mark renderer', () => {
    const options: Options = {
      renderMark: {
        [MARKS.BOLD]: text => <Strong>{text}</Strong>,
      },
    };
    const document: Document = multiMarkDoc();

    expect(documentToReactComponents(document, options)).toMatchSnapshot();
  });

  it('renders text with the passed custom text renderer', () => {
    const options: Options = {
      renderText: text => text.replace(/world/, 'Earth'),
    };
    const document: Document = paragraphDoc;

    expect(documentToReactComponents(document, options)).toMatchSnapshot();
  });

  it('does not render unrecognized marks', () => {
    const document: Document = invalidMarksDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders empty node if type is not recognized', () => {
    const document: Document = invalidTypeDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders default entry link block', () => {
    const entrySys = {
      sys: {
        id: '9mpxT4zsRi6Iwukey8KeM',
        link: 'Link',
        linkType: 'Entry',
      },
    };
    const document: Document = embeddedEntryDoc(entrySys);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders ordered lists', () => {
    const document: Document = olDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders unordered lists', () => {
    const document: Document = ulDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders blockquotes', () => {
    const document: Document = quoteDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders horizontal rule', () => {
    const document: Document = hrDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('does not crash with inline elements (e.g. hyperlink)', () => {
    const document: Document = hyperlinkDoc;

    expect(documentToReactComponents(document)).toBeTruthy();
  });

  it('renders hyperlink', () => {
    const document: Document = hyperlinkDoc;

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });

  it('renders asset hyperlink', () => {
    const asset = {
      target: {
        sys: {
          id: '9mpxT4zsRi6Iwukey8KeM',
          link: 'Link',
          type: 'Asset',
        },
      },
    };
    const document: Document = inlineEntityDoc(asset, INLINES.ASSET_HYPERLINK);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });
  it('renders entry hyperlink', () => {
    const entry = {
      target: {
        sys: {
          id: '9mpxT4zsRi6Iwukey8KeM',
          link: 'Link',
          type: 'Entry',
        },
      },
    };
    const document: Document = inlineEntityDoc(entry, INLINES.ENTRY_HYPERLINK);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });
  it('renders embedded entry', () => {
    const entry = {
      target: {
        sys: {
          id: '9mpxT4zsRi6Iwukey8KeM',
          link: 'Link',
          type: 'Entry',
        },
      },
    };
    const document: Document = inlineEntityDoc(entry, INLINES.EMBEDDED_ENTRY);

    expect(documentToReactComponents(document)).toMatchSnapshot();
  });
});

describe('appendKeyToValidElement', () => {
  it('appends keys to default React components', () => {
    expect(appendKeyToValidElement(<div />, 0)).toHaveProperty('key', '0');
  });

  it('appends keys to custom React components', () => {
    expect(appendKeyToValidElement(<Paragraph>hello world</Paragraph>, 0)).toHaveProperty(
      'key',
      '0',
    );
  });

  it('does not overwrite user specified keys', () => {
    expect(appendKeyToValidElement(<div key={'xyz'} />, 0)).toHaveProperty('key', 'xyz');
  });

  it('does not add keys to text nodes', () => {
    expect(appendKeyToValidElement('hello world', 0)).not.toHaveProperty('key');
  });

  it('does not add keys to node arrays', () => {
    expect(appendKeyToValidElement([<div key={0} />, <div key={1} />], 0)).not.toHaveProperty(
      'key',
    );
  });

  it('does not add keys to null', () => {
    expect(appendKeyToValidElement(null, 0)).toBeNull();
  });
});

describe('nodeToReactComponent', () => {
  const options: Options = {
    renderNode: {
      [BLOCKS.PARAGRAPH]: (node: CommonNode, children: ReactNode): ReactNode => <p>{children}</p>,
    },
    renderMark: {
      [MARKS.BOLD]: (text: ReactNode): ReactNode => <b>{text}</b>,
    },
  };

  const createBlockNode = (nodeType: BLOCKS): CommonNode => ({
    nodeType,
    data: {},
    content: [
      {
        nodeType: 'text',
        value: 'hello world',
        marks: [],
        data: {},
      },
    ],
  });

  const createTextNode = (type: string): CommonNode => ({
    nodeType: 'text',
    value: 'hello world',
    marks: [{ type }],
    data: {},
  });

  it('renders valid nodes', () => {
    expect(nodeToReactComponent(createBlockNode(BLOCKS.PARAGRAPH), options)).toMatchSnapshot();
  });

  it('renders invalid node types in React fragments', () => {
    expect(nodeToReactComponent(createBlockNode(BLOCKS.HEADING_1), options)).toMatchSnapshot();
  });

  it('renders valid marks', () => {
    expect(nodeToReactComponent(createTextNode(MARKS.BOLD), options)).toMatchSnapshot();
  });

  it('does not add additional tags on invalid marks', () => {
    expect(nodeToReactComponent(createTextNode(MARKS.ITALIC), options)).toMatchSnapshot();
  });

  const customTextNode: CommonNode = {
    nodeType: BLOCKS.PARAGRAPH,
    data: {},
    content: [
      {
        nodeType: 'text',
        value: 'some\nlines\nof\ntext',
        marks: [{ type: MARKS.BOLD }],
        data: {},
      },
    ],
  };

  it('does not render altered text with default text renderer', () => {
    expect(nodeToReactComponent(customTextNode, options)).toMatchSnapshot();
  });

  it('renders altered text with custom text renderer', () => {
    expect(
      nodeToReactComponent(customTextNode, {
        ...options,
        renderText: (text: string): ReactNode => {
          return text.split('\n').reduce((children, textSegment, index) => {
            return [...children, index > 0 && <br key={index} />, textSegment];
          }, []);
        },
      }),
    ).toMatchSnapshot();
  });
});

describe('nodeListToReactComponents', () => {
  const options: Options = {
    renderNode: {
      [BLOCKS.PARAGRAPH]: (node: CommonNode, children: ReactNode): ReactNode => <p>{children}</p>,
    },
    renderMark: {
      [MARKS.BOLD]: (text: ReactNode): ReactNode => <b>{text}</b>,
    },
  };

  const nodes: CommonNode[] = [
    {
      nodeType: BLOCKS.PARAGRAPH,
      data: {},
      content: [
        {
          nodeType: 'text',
          value: 'hello',
          marks: [{ type: MARKS.BOLD }],
          data: {},
        },
      ],
    },
    {
      nodeType: 'text',
      value: ' ',
      marks: [],
      data: {},
    },
    {
      nodeType: BLOCKS.PARAGRAPH,
      data: {},
      content: [
        {
          nodeType: 'text',
          value: 'world',
          marks: [],
          data: {},
        },
      ],
    },
  ];

  it('renders children as an array with keys from its index', () => {
    const renderedNodes: ReactNodeArray = nodeListToReactComponents(
      nodes,
      options,
    ) as ReactNodeArray;
    expect(renderedNodes[0]).toHaveProperty('key', '0');
    expect(renderedNodes[1]).not.toHaveProperty('key');
    expect(renderedNodes[2]).toHaveProperty('key', '2');
    expect(renderedNodes).toMatchSnapshot();
  });
});
