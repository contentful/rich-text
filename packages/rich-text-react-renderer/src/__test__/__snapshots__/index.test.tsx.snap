// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentToReactComponents does not render unrecognized marks 1`] = `
Array [
  <p>
    Hello world!
  </p>,
]
`;

exports[`documentToReactComponents renders asset hyperlink 1`] = `
Array [
  <p>
    
    <span>
      type: 
      asset-hyperlink
       id: 
      9mpxT4zsRi6Iwukey8KeM
    </span>
    
  </p>,
]
`;

exports[`documentToReactComponents renders blockquotes 1`] = `
Array [
  <p>
    hello
  </p>,
  <blockquote>
    world
  </blockquote>,
]
`;

exports[`documentToReactComponents renders default entry link block 1`] = `
Array [
  <div />,
]
`;

exports[`documentToReactComponents renders embedded entry 1`] = `
Array [
  <p>
    
    <span>
      type: 
      embedded-entry-inline
       id: 
      9mpxT4zsRi6Iwukey8KeM
    </span>
    
  </p>,
]
`;

exports[`documentToReactComponents renders empty node if type is not recognized 1`] = `
Array [
  <React.Fragment>
    Hello world!
  </React.Fragment>,
]
`;

exports[`documentToReactComponents renders entry hyperlink 1`] = `
Array [
  <p>
    
    <span>
      type: 
      entry-hyperlink
       id: 
      9mpxT4zsRi6Iwukey8KeM
    </span>
    
  </p>,
]
`;

exports[`documentToReactComponents renders horizontal rule 1`] = `
Array [
  <p>
    hello world
  </p>,
  <hr />,
  <p>
    
  </p>,
]
`;

exports[`documentToReactComponents renders hyperlink 1`] = `
Array [
  <p>
    Some text 
    <a
      href="https://url.org"
    >
      link
    </a>
     text.
  </p>,
]
`;

exports[`documentToReactComponents renders marks with default mark renderer 1`] = `
Array [
  <p>
    <i>
      hello world
    </i>
  </p>,
]
`;

exports[`documentToReactComponents renders marks with default mark renderer 2`] = `
Array [
  <p>
    <b>
      hello world
    </b>
  </p>,
]
`;

exports[`documentToReactComponents renders marks with default mark renderer 3`] = `
Array [
  <p>
    <u>
      hello world
    </u>
  </p>,
]
`;

exports[`documentToReactComponents renders marks with default mark renderer 4`] = `
Array [
  <p>
    <code>
      hello world
    </code>
  </p>,
]
`;

exports[`documentToReactComponents renders marks with the passed custom mark renderer 1`] = `
Array [
  <p>
    <i>
      <Strong>
        hello world
      </Strong>
    </i>
  </p>,
]
`;

exports[`documentToReactComponents renders multiple marks with default mark renderer 1`] = `
Array [
  <p>
    <i>
      <b>
        hello world
      </b>
    </i>
  </p>,
]
`;

exports[`documentToReactComponents renders nodes with default node renderer 1`] = `
Array [
  <p>
    hello world
  </p>,
]
`;

exports[`documentToReactComponents renders nodes with default node renderer 2`] = `
Array [
  <h1>
    hello world
  </h1>,
]
`;

exports[`documentToReactComponents renders nodes with default node renderer 3`] = `
Array [
  <h2>
    hello world
  </h2>,
]
`;

exports[`documentToReactComponents renders nodes with passed custom node renderer 1`] = `
Array [
  <Paragraph>
    hello
  </Paragraph>,
  <blockquote>
    world
  </blockquote>,
]
`;

exports[`documentToReactComponents renders ordered lists 1`] = `
Array [
  <ol>
    <li>
      <p>
        Hello
      </p>
    </li>
    <li>
      <p>
        world
      </p>
    </li>
  </ol>,
  <p>
    
  </p>,
]
`;

exports[`documentToReactComponents renders text with the passed custom text renderer 1`] = `
Array [
  <p>
    hello Earth
  </p>,
]
`;

exports[`documentToReactComponents renders unaltered text with default text renderer 1`] = `
Array [
  <p>
    hello world
  </p>,
]
`;

exports[`documentToReactComponents renders unordered lists 1`] = `
Array [
  <ul>
    <li>
      <p>
        Hello
      </p>
    </li>
    <li>
      <p>
        world
      </p>
    </li>
  </ul>,
  <p>
    
  </p>,
]
`;

exports[`nodeListToReactComponents renders children as an array with keys from its index 1`] = `
Array [
  <p>
    <b>
      hello
    </b>
  </p>,
  " ",
  <p>
    world
  </p>,
]
`;

exports[`nodeToReactComponent does not add additional tags on invalid marks 1`] = `"hello world"`;

exports[`nodeToReactComponent does not render altered text with default text renderer 1`] = `
<p>
  <b>
    some
lines
of
text
  </b>
</p>
`;

exports[`nodeToReactComponent renders altered text with custom text renderer 1`] = `
<p>
  <b>
    some
    <br />
    lines
    <br />
    of
    <br />
    text
  </b>
</p>
`;

exports[`nodeToReactComponent renders invalid node types in React fragments 1`] = `
<React.Fragment>
  hello world
</React.Fragment>
`;

exports[`nodeToReactComponent renders valid marks 1`] = `
<b>
  hello world
</b>
`;

exports[`nodeToReactComponent renders valid nodes 1`] = `
<p>
  hello world
</p>
`;
